src/openrct2-ui/input/InputManager.h

```diff
@@ -0,0 +1,83 @@
+/*****************************************************************************
+ * Copyright (c) 2014-2020 OpenRCT2 developers
+ *
+ * For a complete list of all authors, please refer to contributors.md
+ * Interested in contributing? Visit https://github.com/OpenRCT2/OpenRCT2
+ *
+ * OpenRCT2 is licensed under the GNU General Public License version 3.
+ *****************************************************************************/
+
+#pragma once
+
+#include <openrct2/world/Location.hpp>
+#include <queue>
+#include <string_view>
+
+typedef struct _SDL_Joystick SDL_Joystick;
+typedef union SDL_Event SDL_Event;
+
+namespace OpenRCT2::Ui
+{
+    class RegisteredShortcut;
+    struct ShortcutInput;
+
+    enum class InputDeviceKind
+    {
+        Mouse,
+        Keyboard,
+        JoyButton,
+        JoyHat,
+    };
+
+    enum class InputEventState
+    {
+        Down,
+        Release,
+    };
+
+    struct InputEvent
+    {
+        InputDeviceKind DeviceKind;
+        uint32_t Modifiers;
+        uint32_t Button;
+        InputEventState State;
+    };
+
+    class InputManager
+    {
+    private:
+        uint32_t _lastJoystickCheck;
+        std::vector<SDL_Joystick*> _joysticks;
+        std::queue<InputEvent> _events;
+        ScreenCoordsXY _viewScroll;
+        uint32_t _mouseState;
+        std::vector<uint8_t> _keyboardState;
+
+        RegisteredShortcut* _scrollLeftShortcut{};
+        RegisteredShortcut* _scrollUpShortcut{};
+        RegisteredShortcut* _scrollRightShortcut{};
+        RegisteredShortcut* _scrollDownShortcut{};
+
+        void CheckJoysticks();
+
+        void HandleViewScrolling();
+        void HandleModifiers();
+        void ProcessEvents();
+        void Process(const InputEvent& e);
+        void ProcessInGameConsole(const InputEvent& e);
+        void ProcessChat(const InputEvent& e);
+        void ProcessHoldEvents();
+        void ProcessViewScrollEvent(std::string_view shortcutId, RegisteredShortcut*& shortcut, const ScreenCoordsXY& delta);
+
+        bool GetState(const RegisteredShortcut& shortcut) const;
+        bool GetState(const ShortcutInput& shortcut) const;
+
+    public:
+        void QueueInputEvent(const SDL_Event& e);
+        void QueueInputEvent(InputEvent&& e);
+        void Process();
+    };
+
+    InputManager& GetInputManager();
```

Comment:
Shouldn't this be in the Context?

Reply:
Context or UiContext?

Reply:
Context

Reply:
InputManager is currently in libopenrct2ui, so I can't move it to Context. It doesn't make sense to either as any core code shouldn't be calling anything to do with input.

Reply:
Technically the interactive console requires input but I see that we use Linenoise so in this case it doesn't matter I guess, so UiContext it is.

Reply:
ok done